### PR TITLE
Take control of view lifecycle events

### PR DESCRIPTION
**Pull request by vinaysshenoy**
https://www.pivotaltracker.com/story/show/166784013
https://www.pivotaltracker.com/story/show/166783916

### CODE REVIEW

File: app/src/main/java/org/simple/clinic/ViewControllerBinding.kt [hunk 1]
```diff
@@ -22,12 +22,13 @@ fun <T> bindUiToController(
       .mergeWith(screenDestroys)
       .observeOn(io())
       .compose(controller)
-      .delaySubscription(uiChangeDelay.toMillis(), TimeUnit.MILLISECONDS)
+      .delaySubscription(uiChangeDelay.toMillis(), TimeUnit.MILLISECONDS, mainThread())
```
Comment (saket) [inline app/src/main/java/org/simple/clinic/ViewControllerBinding.kt]: Curious question: what's the use of `delaySubscription()` in this context?

File: app/src/main/java/org/simple/clinic/allpatientsinfacility/AllPatientsInFacilityView.kt [hunk 2]
```diff
@@ -117,10 +122,22 @@ class AllPatientsInFacilityView(
   }
 
   override fun onSaveInstanceState(): Parcelable? {
-    return super.onSaveInstanceState()
+    return Bundle().apply {
+      val viewState = super.onSaveInstanceState()
+      putParcelable(VIEW_STATE_KEY, viewState)
+      putParcelable(FEATURE_STATE_KEY, states.value)
+    }
```
Comment (saket) [inline app/src/main/java/org/simple/clinic/allpatientsinfacility/AllPatientsInFacilityView.kt]: Will `ScreenSavedState` be useful here?

File: app/src/main/java/org/simple/clinic/ViewControllerBinding.kt (hunk 1 above)
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/ViewControllerBinding.kt]: That's to subscribe to the view events on the main thread. With the changes to the arch we are running, subscriptions started happening on the UI thread, and RxBinding started complaining and crashing. 💣

File: app/src/main/java/org/simple/clinic/allpatientsinfacility/AllPatientsInFacilityView.kt (hunk 2 above)
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/allpatientsinfacility/AllPatientsInFacilityView.kt]: We are exploring different options.

File: app/src/main/java/org/simple/clinic/ViewControllerBinding.kt (hunk 1 above)
Comment (saket) [reply app/src/main/java/org/simple/clinic/ViewControllerBinding.kt]: Got it

**Comment by vinaysshenoy** [comment]
@ragunathjawahar I have rebased this branch on master.

**Comment by vinaysshenoy** [comment]
Set to WIP until #747 is done since the last commit will depend on this.

File: app/src/main/java/org/simple/clinic/ViewControllerBinding.kt [hunk 3]
```diff
@@ -44,3 +58,63 @@ fun <T> bindUiToControllerWithoutDelay(
       .takeUntil(screenDestroys)
       .subscribe { uiChange -> uiChange(ui) }
 }
+
+class ViewControllerBinding<E, S, T>(
```
Comment (ragunathjawahar) [inline app/src/main/java/org/simple/clinic/ViewControllerBinding.kt]: `<E, S, U>` or `<E, S, V>`?

**Pull request merged**